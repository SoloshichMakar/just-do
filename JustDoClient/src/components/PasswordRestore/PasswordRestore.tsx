import React from "react";
import { Snackbar } from "@material-ui/core";
import JustDoLogo from "../../picture/JustDoLogo.png";
import JustDoComponent from "../../picture/JustDoComponent.png";
import BackArrow from "../../picture/ic_arrow_left.svg";
import FormControl from "@material-ui/core/FormControl";
import Input from "@material-ui/core/Input";
import { Link, useHistory } from "react-router-dom";
import {
  confirmPasswordHelperMessage,
  passwordHelperMessage,
  validateConfirmPassword,
  validatePassword,
} from "../../utils/validateUserDataFunction";
import { gql } from "apollo-boost";
import { useMutation } from "@apollo/react-hooks";
import { Alert } from "@material-ui/lab";
import { ColorButton } from "../style/CustomElements";

interface IForgotPasswordRestore {
  password: string;
  message: string;
  confirmPassword: string;
  match: any;
  setMessage: Function;
  errorMessage: Function;
  clearMessage: Function;
  textPasswordChange: Function;
  textConfirmPasswordChange: Function;
}

const PASSWORD_RESTORE = gql`
  mutation PasswordRestore($id: ID!, $password: String!, $token: String!) {
    passwordRestore(id: $id, password: $password, token: $token) {
      id
      email
    }
  }
`;

let passwordIsEdit = false;
let confirmPasswordIsEdit = false;

const PasswordRestore: React.FC<IForgotPasswordRestore> = ({
  password,
  confirmPassword,
  message,
  errorMessage,
  clearMessage,
  setMessage,
  textPasswordChange,
  textConfirmPasswordChange,
  match,
}) => {
  const [passwordRestore] = useMutation(PASSWORD_RESTORE);
  const history = useHistory();

  function currentTextPasswordChange(e: React.ChangeEvent<HTMLInputElement>) {
    passwordIsEdit = true;
    const getTextAreaValue = e.target.value;
    textPasswordChange(getTextAreaValue);
  }

  function currentTextConfirmPasswordChange(
    e: React.ChangeEvent<HTMLInputElement>
  ) {
    confirmPasswordIsEdit = true;
    const getTextAreaValue = e.target.value;
    textConfirmPasswordChange(getTextAreaValue);
  }

  async function updatePassword() {
    try {
      if (
        !validatePassword(passwordIsEdit, password) &&
        passwordIsEdit &&
        !validateConfirmPassword(
          confirmPasswordIsEdit,
          password,
          confirmPassword
        ) &&
        confirmPasswordIsEdit
      ) {
        await passwordRestore({
          variables: {
            id: parseInt(match.params.id),
            password: password,
            token: match.params.token,
          },
        });
        passwordIsEdit = false;
        confirmPasswordIsEdit = false;
        setMessage("You have successfully changed your password!");
        history.push("/login");
      }
    } catch (error) {
      errorMessage();
    }
  }

  const handleErrorClose = (event?: React.SyntheticEvent, reason?: string) => {
    if (reason === "clickaway") {
      return;
    }
    clearMessage();
  };

  return (
    <div id="forgot_password_main">
      <Snackbar
        open={message.length !== 0}
        autoHideDuration={10000}
        onClose={handleErrorClose}
      >
        <Alert onClose={handleErrorClose} severity="error">
          {message}
        </Alert>
      </Snackbar>
      <div id="forgot_password_main__logo_images">
        <img id="logo" src={JustDoLogo} />
        <img src={JustDoComponent} id="background_image" />
      </div>
      <div id="forgot_password_main__components">
        <Link id="arrow_link" to="/login">
          <img src={BackArrow} />
        </Link>
        <h1>Password restore</h1>
        <div id="components_helper_text">
          {" "}
          Please enter and confirm your new password{" "}
        </div>
        <div id="components__input_components">
          <div className="input_components__text_input">
            <FormControl id="textInput__passwordForm">
              <Input
                id="password_input"
                error={validatePassword(passwordIsEdit, password)}
                type={"password"}
                value={password}
                placeholder="Password"
                onChange={currentTextPasswordChange}
              />
              <div id="errorMessage">
                {passwordHelperMessage(passwordIsEdit, password)}
              </div>
            </FormControl>

            <div className="input_components__text_input">
              <FormControl id="textInput__passwordConfirmForm">
                <Input
                  id="password_confirm_input"
                  error={validateConfirmPassword(
                    passwordIsEdit,
                    password,
                    confirmPassword
                  )}
                  type="password"
                  value={confirmPassword}
                  onChange={currentTextConfirmPasswordChange}
                  placeholder="Confirm password"
                />

                <div id="errorMessage">
                  {confirmPasswordHelperMessage(
                    confirmPasswordIsEdit,
                    password,
                    confirmPassword
                  )}
                </div>
              </FormControl>
            </div>

            <div id="input_components__color_button">
              <ColorButton
                id="color_button"
                variant="contained"
                color="primary"
                onClick={() => updatePassword()}
              >
                Send
              </ColorButton>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default PasswordRestore;
