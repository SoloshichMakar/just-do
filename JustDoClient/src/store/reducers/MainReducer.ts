import RegistrationReducer from "./RegistrationReducer";
import AuthorizationReducer from "./AuthorizationReducer";
import PasswordRestoreReducer from "./PasswordRestoreReducer";
import ForgotPasswordReducer from "./ForgotPasswordReducer";

import { combineReducers } from "redux";

const MainReducer = combineReducers({
  RegistrationReducer,
  AuthorizationReducer,
  PasswordRestoreReducer,
  ForgotPasswordReducer,
});

export default MainReducer;
