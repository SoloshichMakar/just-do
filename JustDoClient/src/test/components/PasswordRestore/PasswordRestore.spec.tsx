import renderer from "react-test-renderer";
import { createStore } from "redux";
import MainReducer from "../../../store/reducers/MainReducer";
import { Provider } from "react-redux";
import React from "react";
import PasswordRestore from "../../../components/PasswordRestore/PasswordRestore";
import ApolloClient from "apollo-boost";
import { ApolloProvider } from "react-apollo";
import { BrowserRouter } from "react-router-dom";

const props = {
    password: "",
    message: "",
    confirmPassword: "",
    match: Function,
    setMessage: Function,
    errorMessage: Function,
    clearMessage: Function,
    textPasswordChange: Function,
    textConfirmPasswordChange: Function,
};

const client = new ApolloClient({
    uri: ``,
});
const store = createStore(MainReducer);

describe("Password restore component snapshot", () => {
    it("Password restore component render correctly", () => {
        const tree = renderer
            .create(
                <ApolloProvider client={client}>
                    <Provider store={store}>
                        <BrowserRouter>
                            <PasswordRestore
                               password={props.password}
                               message={props.message}
                               confirmPassword={props.confirmPassword}
                               match={props.match}
                               setMessage={props.setMessage}
                               errorMessage={props.errorMessage}
                               clearMessage={props.clearMessage}
                               textPasswordChange={props.textPasswordChange}
                               textConfirmPasswordChange={props.textConfirmPasswordChange}
                            />
                        </BrowserRouter>
                    </Provider>
                </ApolloProvider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});