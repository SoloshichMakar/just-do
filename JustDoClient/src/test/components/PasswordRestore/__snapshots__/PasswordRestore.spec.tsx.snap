// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password restore component snapshot Password restore component render correctly 1`] = `
<div
  id="forgot_password_main"
>
  <div
    id="forgot_password_main__logo_images"
  >
    <img
      id="logo"
      src="JustDoLogo.png"
    />
    <img
      id="background_image"
      src="JustDoComponent.png"
    />
  </div>
  <div
    id="forgot_password_main__components"
  >
    <a
      href="/login"
      id="arrow_link"
      onClick={[Function]}
    >
      <img
        src="ic_arrow_left.svg"
      />
    </a>
    <h1>
      Password restore
    </h1>
    <div
      id="components_helper_text"
    >
       
      Please enter and confirm your new password
       
    </div>
    <div
      id="components__input_components"
    >
      <div
        className="input_components__text_input"
      >
        <div
          className="MuiFormControl-root"
          id="textInput__passwordForm"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="password_input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              required={false}
              type="password"
              value=""
            />
          </div>
          <div
            id="errorMessage"
          />
        </div>
        <div
          className="input_components__text_input"
        >
          <div
            className="MuiFormControl-root"
            id="textInput__passwordConfirmForm"
          >
            <div
              className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                className="MuiInputBase-input MuiInput-input"
                disabled={false}
                id="password_confirm_input"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Confirm password"
                required={false}
                type="password"
                value=""
              />
            </div>
            <div
              id="errorMessage"
            />
          </div>
        </div>
        <div
          id="input_components__color_button"
        >
          <button
            className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-44 MuiButton-contained MuiButton-containedPrimary"
            disabled={false}
            id="color_button"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Send
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Registration snapshot Registration render correctly 1`] = `
<div
  id="forgot_password_main"
>
  <div
    id="forgot_password_main__logo_images"
  >
    <img
      id="logo"
      src="JustDoLogo.png"
    />
    <img
      id="background_image"
      src="JustDoComponent.png"
    />
  </div>
  <div
    id="forgot_password_main__components"
  >
    <a
      href="/login"
      id="arrow_link"
      onClick={[Function]}
    >
      <img
        src="ic_arrow_left.svg"
      />
    </a>
    <h1>
      Password restore
    </h1>
    <div
      id="components_helper_text"
    >
       
      Please enter and confirm your new password
       
    </div>
    <div
      id="components__input_components"
    >
      <div
        className="input_components__text_input"
      >
        <div
          className="MuiFormControl-root"
          id="textInput__passwordForm"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="password_input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              required={false}
              type="password"
              value=""
            />
          </div>
          <div
            id="errorMessage"
          />
        </div>
        <div
          className="input_components__text_input"
        >
          <div
            className="MuiFormControl-root"
            id="textInput__passwordConfirmForm"
          >
            <div
              className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                className="MuiInputBase-input MuiInput-input"
                disabled={false}
                id="password_confirm_input"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Confirm password"
                required={false}
                type="password"
                value=""
              />
            </div>
            <div
              id="errorMessage"
            />
          </div>
        </div>
        <div
          id="input_components__color_button"
        >
          <button
            className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-44 MuiButton-contained MuiButton-containedPrimary"
            disabled={false}
            id="color_button"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Send
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
